### ISSUE TITLE
Tracking issue for #[wasm_custom_section] and #[wasm_import_module] attributes

### ISSUE THREAD

**Issue by kennytm** [rust-lang/rust#51088]
This tracks the `wasm_custom_section` and `wasm_import_module` features which introduced two same-named attributes.

cc rust-lang-nursery/rust-wasm#82.

```rust
#[wasm_custom_section = "foo"]
const CUSTOM_SECTION_CONTENT: [u8; 5] = [1, 2, 3, 4, 5];

#[wasm_import_module = "./me"]
extern {
    #[link_name = "me_in_dep"]
    fn dep();
}
```

- [x] Implementation (#48883)
- [ ] Stabilization

**Comment by alexcrichton** [comment]
I've commandeered this issue for just `#[wasm_custom_section]` and moved `#[wasm_import_module]` to a [different issue](https://github.com/rust-lang/rust/issues/52090) as I suspect this will be more controversial than the import modules.

**Comment by kennytm** [comment]
Oh, `feature_gate.rs` will need to be updated to the new tracking issue (#52090).

https://github.com/rust-lang/rust/blob/00f49728e03d7bcf419e0d94fa45f3f4285ff7e3/src/libsyntax/feature_gate.rs#L423-L424

**Comment by alexcrichton** [comment]
Similar to https://github.com/rust-lang/rust/issues/52090 I think this feature is ready enough for stabilization now, could I perhaps persuade a member of @rust-lang/compiler to start the FCP for this issue? There's a description of how this issue works [in the OP](https://github.com/rust-lang/rust/issues/51088).

Additionally, cc @rust-lang/wg-wasm, y'all are probably interested too!

For me the motivation for this issue is to ensure that `wasm-bindgen` works on stable for the 2018 edition.

**Comment by kennytm** [comment]
Questions:

1. Why a `const` instead of a `static`
2. Why a new attribute instead of reusing `#[link_section]`

**Comment by oli-obk** [comment]
> The linker, LLD, for the wasm target will concatenate custom sections of the same name in the order the objects appear on the comand line. For example if both crate a and crate b define a custom section with the name foo the final output will contain only one custom section with the name of foo where the contents of a and b are concatenated (byte-wise).

If the same section name is used multiple times within a crate, rustc currently accepts this (http://play.rust-lang.org/?gist=d586d1b74b220b29579472bca8f4167d&version=nightly&mode=debug&edition=2015) but I'm fairly certain that this causes an undefined ordering. Especially with incremental involved. Even if it doesn't, reordering modules with sections inside will cause these sections to be reordered.

I'm assuming we could just forbid using the same section name twice within a single crate?

**Comment by alexcrichton** [comment]
@kennytm 

> Why a `const` instead of a `static`

The general idea I had here was that `static` represents an addressable value in memory, but here custom sections aren't at all accessible by normal code, so I somewhat arbitrarily chose `const`.

> Why a new attribute instead of reusing `#[link_section]`

A good question! If we went with `#[link_section]` we wouldn't even have to FCP anything :). I originally did a custom attribute due to the difference with `#[link_section]` but it may be best to just go back and use it (I'd be up for either). The `#[link_section]` attribute is more applicable to other platforms like ELF and COFF whereas for wasm you can't shove arbitrary data and constants into custom sections. Additionally you can't have things like relocations to those sections.

So originally I chose a different attribute because it could (a) be unstable vs `#[link_section]`, (b) didn't want to conflate it with being attached to anything, and (c) had strict requirements (currently) on the form of only allowing `[u8; N]`. Now whether these are good ideas to stick with remains an unknown!

---

@oli-obk 

Correct it is accepted multiple times within one crate! Also correct that there's a somewhat undefined ordering coming out the other end. I don't think the problem is exclusively within one crate though in that you can't really say much about orderings between crates either. In general I believe this just means that you either have to design custom sections to be unique or binary-concatenatable-yet-still-understandable.

Another motivation for supporting multiple uses within a crate is that each `#[wasm_bindgen]` macro generates a custom section entry, so not allowing it there would be a problem!

**Comment by oli-obk** [comment]
@rfcbot fcp merge

**Comment by eddyb** [comment]
I prefer `#[link_section]` as well. Note that the format is already platform-dependent, i.e. for macOS' Mach-O you need something like `__DATA,__data` instead of `.data`.

**Comment by eddyb** [comment]
@alexcrichton wait, why the byte restriction? You can get the byte contents of any `miri` allocation, the only restriction you have to impose is no relocations (and other executable formats also have relocation restrictions).

**Comment by alexcrichton** [comment]
@eddyb Ok cool! That may be possible yeah. In that sense this could also be an issue in that case to change to:

* Remove `#[wasm_custom_section]` in favor of `#[link_section]`
* On the wasm target, `#[link_section]` can only be applied to `static` values, erroring out on all other locations
* On the wasm target, `#[link_section]` requires no relocations 

Denying relocations means denying this, right?

```rust
#[link_section = "foo"]
static A: &u8 = &1;
```

**Comment by eddyb** [comment]
@alexcrichton Yeah. `miri` literally uses a relocation model, so it should be easy to implement.

**Comment by alexcrichton** [comment]
@eddyb hm but it still leaves a weird question I think in terms of:

```rust
#[link_section = "foo"]
static A: u8 = 8;

fn main() {
    println!("{}", A);
}
```

This means that `A` is *both* in the main memory as well as a custom section? On all other platforms main code can reference custom sections?

**Comment by eddyb** [comment]
@alexcrichton IIRC at least in ELF sections can be marked as not being loaded to memory at all, I wonder what happens then! My guess is either the relocation in `A` fails to link or the code segfaults.

**Comment by alexcrichton** [comment]
Ok I've submitted https://github.com/rust-lang/rust/pull/52353 with @eddyb's suggestion, and if that's merged it'll just close this issue
